packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -170,109 +157,106 @@ export default function parse(source: string, state: State): TemplateParseResult
-        Element: {
-            enter(node) {
-                const elementNode = node as parse5.AST.Default.Element;
+    const root = parseElement(templateRoot);
 
-                const element = createElement(node);
-                element.attrsList = elementNode.attrs;
+    function parseElement(elementNode: parse5.AST.Default.Element, parent?: IRElement): IRElement {
+        const element = createElement(elementNode, parent);
 
-                if (!root) {
-                    root = element;
-                } else {
-                    element.parent = parent;
-                    parent.children.push(element);
-                }
+        applyForEach(element);
+        applyIterator(element);
+        applyIf(element);
+        applyHandlers(element);
+        applyComponent(element);
+        applySlot(element);
+        applyKey(element);
+        applyLwcDirectives(element);
+        applyAttributes(element);
+        validateElement(element);
+        validateAttributes(element);
+        validateProperties(element);
+
+        parseChildren(element);
+        validateChildren(element);
+
+        return element;
+    }
+
+    function parseChildren(parent: IRElement): void {
+        const { __original } = parent;
+
+        const parsedChildren: IRNode[] = [];
+        const children = treeAdapter.getChildNodes(
+            treeAdapter.getTemplateContent(__original) ?? __original
+        );
 
-                validateInlineStyleElement(element);
-
-                applyForEach(element);
-                applyIterator(element);
-                applyIf(element);
-                applyHandlers(element);
-                applyComponent(element);
-                applySlot(element);
-                applyKey(element, elementNode.__location);
-
-                parent = element;
-                stack.push(element);
-            },
-            exit() {
-                const element = stack.pop() as IRElement;
-                // Apply lwc directive on way out to ensure the element is empty
-                applyLwcDirectives(element);
-                applyAttributes(element);
-                validateElement(element);
-                validateAttributes(element);
-                validateProperties(element);
-
-                parent = stack[stack.length - 1];
-            },
-        },
-
-        Text: {
-            enter(node: parse5.AST.Default.TextNode) {
-                // Extract the raw source to avoid HTML entity decoding done by parse5
-                // TODO [#1286]: Update parse5-with-error to match version used for jsdom (interface for ElementLocation changed)
-                const location = node.__location as parse5.MarkupData.Location;
-
-                const { startOffset, endOffset } = location;
-                const rawText = cleanTextNode(source.slice(startOffset, endOffset));
-
-                if (!rawText.trim().length) {
-                    return;
+        if (children !== undefined) {
```

Comment:
This if check is unnecessary. The return type for getChildNodes is `AST.Default.Node[]`.